File: app/src/main/java/org/freelesson/pagingsample/Injection.java
package org.freelesson.pagingsample;

import android.content.Context;

import androidx.lifecycle.ViewModelProvider;

import org.freelesson.pagingsample.api.GithubService;
import org.freelesson.pagingsample.data.GithubRepository;
import org.freelesson.pagingsample.db.GithubLocalCache;
import org.freelesson.pagingsample.db.RepoDatabase;
import org.freelesson.pagingsample.ui.ViewModelFactory;

import java.util.concurrent.Executors;

public class Injection {
    private  static GithubLocalCache provideCache(Context context) {
        RepoDatabase database = RepoDatabase.getInstance(context);
        return new GithubLocalCache(database.repoDaos(), Executors.newSingleThreadExecutor());
    }
    private static GithubRepository provideGithubRepository(Context context) {
        return new GithubRepository(GithubService.create(),provideCache(context));
    }
    public static ViewModelProvider.Factory provideViewModelFactory(Context context) {
        return new ViewModelFactory(provideGithubRepository(context));
    }
}

File: app/src/main/java/org/freelesson/pagingsample/Injection.kt
package org.freelesson.pagingsample

import android.content.Context
import androidx.lifecycle.ViewModelProvider
import androidx.savedstate.SavedStateRegistryOwner
import org.freelesson.pagingsample.api.GithubService
import org.freelesson.pagingsample.data.GithubRepository
import org.freelesson.pagingsample.db.RepoDatabase
import org.freelesson.pagingsample.ui.ViewModelFactory

object Injection {
    private fun provideGithubRepository(context: Context): GithubRepository {
        return GithubRepository(GithubService.create(), RepoDatabase.getInstance(context))
    }

    fun provideViewModelFactory(
        context: Context,
        owner: SavedStateRegistryOwner
    ): ViewModelProvider.Factory {
        return ViewModelFactory(owner, provideGithubRepository(context))
    }
}